File: anvil-api/src/main/java/org/anvilpowered/anvil/api/misc/BindingExtensions.java
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020-2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.api.misc;

import com.google.common.reflect.TypeToken;
import com.google.inject.Key;
import com.google.inject.Provider;
import com.google.inject.TypeLiteral;
import org.anvilpowered.anvil.api.Anvil;
import org.anvilpowered.anvil.api.datastore.DBComponent;
import org.anvilpowered.anvil.api.datastore.DataStoreContext;

import java.lang.annotation.Annotation;

@SuppressWarnings({"unchecked", "UnstableApiUsage"})
public interface BindingExtensions {

    /**
     * Full binding method for a component
     * <p>
     * A typical example of usage of this method:
     * </p>
     * <pre>{@code
     * be.bind(
     *     new TypeToken<FooRepository<?, ?, ?>>(getClass()) {
     *     },
     *     new TypeToken<FooRepository<?, Foo<?, ?>>(getClass()) {
     *     },
     *     new TypeToken<FooRepository<ObjectId, Foo<ObjectId>, Datastore>>(getClass()) {
     *     },
     *     new TypeToken<CommonMongoFooRepository<TMongoFoo>>(getClass()) { // final implementation
     *     },
     *     Names.named("mongodb")
     * );
     * }</pre>
     */
    <From1 extends DBComponent<?, ?>,
        From2 extends DBComponent<?, ?>,
        From3 extends From1,
        Target extends From1>
    void bind(
        TypeToken<From1> from1,
        TypeToken<From2> from2,
        TypeToken<From3> from3,
        TypeToken<Target> target,
        Annotation componentAnnotation
    );

    /**
     * Binding method for a component
     * <p>
     * A typical example of usage of this method:
     * </p>
     * <pre>{@code
     * be.bind(
     *     new TypeToken<FooRepository<?, ?>>(getClass()) {
     *     },
     *     new TypeToken<FooRepository<ObjectId, Datastore>>(getClass()) {
     *     },
     *     new TypeToken<CommonMongoFooRepository>(getClass()) { // final implementation
     *     },
     *     Names.named("mongodb")
     * );
     * }</pre>
     */
    <From1 extends DBComponent<?, ?>,
        From2 extends From1,
        Target extends From1>
    void bind(
        TypeToken<From1> from1,
        TypeToken<From2> from2,
        TypeToken<Target> target,
        Annotation componentAnnotation
    );

    <From, Target extends From> void bind(
        TypeToken<From> from,
        TypeToken<Target> target,
        Annotation annotation
    );

    <From, Target extends From> void bind(
        TypeToken<From> from,
        TypeToken<Target> target,
        Class<? extends Annotation> annotation
    );

    <From, Target extends From> void bind(
        TypeToken<From> from,
        TypeToken<Target> target
    );

    /**
     * Binds the mongodb {@link DataStoreContext}
     * <p>
     * Using this method is the same as invoking:
     * </p>
     * <pre>{@code
     * bind(new TypeLiteral<DataStoreContext<ObjectId, Datastore>>() {
     * }).to(MongoContext.class);
     * }</pre>
     */
    void withMongoDB();

    /**
     * Binds the xodus {@link DataStoreContext}
     * <p>
     * Using this method is the same as invoking:
     * </p>
     * <pre>{@code
     * binder.bind(new TypeLiteral<DataStoreContext<EntityId, PersistentEntityStore>>() {
     * }).to(XodusContext.class);
     * }</pre>
     */
    void withXodus();


    static <T> TypeLiteral<T> getTypeLiteral(TypeToken<T> typeToken) {
        return (TypeLiteral<T>) TypeLiteral.get(typeToken.getType());
    }

    static <T> Key<T> getKey(TypeToken<T> typeToken) {
        return Key.get(getTypeLiteral(typeToken));
    }

    static <T> Provider<T> asInternalProvider(Class<T> clazz) {
        return () -> Anvil.getEnvironment().getInjector()
            .getInstance(clazz);
    }

    static <T> Provider<T> asInternalProvider(TypeLiteral<T> typeLiteral) {
        return () -> Anvil.getEnvironment().getInjector()
            .getInstance(Key.get(typeLiteral));
    }

    static <T> Provider<T> asInternalProvider(TypeToken<T> typeToken) {
        return () -> Anvil.getEnvironment().getInjector()
            .getInstance(BindingExtensions.getKey(typeToken));
    }
}

File: anvil-api/src/main/kotlin/org/anvilpowered/anvil/api/misc/BindingExtensions.kt
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020-2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.api.misc

import com.google.common.reflect.TypeToken
import com.google.inject.Key
import com.google.inject.Provider
import com.google.inject.TypeLiteral
import org.anvilpowered.anvil.api.Anvil
import org.anvilpowered.anvil.api.datastore.DBComponent

@Suppress("unchecked", "UnstableApiUsage")
interface BindingExtensions {

  /**
   * Full binding method for a component
   *
   * A typical example of usage of this method:
   *
   * be.bind(
   *
   * &nbsp;&nbsp;new TypeToken<FooRepository<?, ?, ?>>(getClass()) {
   * },
   *
   * &nbsp;&nbsp;new TypeToken<FooRepository<?, Foo<?, ?>>(getClass()) {
   * },
   *
   * &nbsp;&nbsp;new TypeToken<FooRepository<ObjectId, Foo<ObjectId>, Datastore>>(getClass()) {
   * },
   *
   * &nbsp;&nbsp;new TypeToken<CommonMongoFooRepository<TMongoFoo>>(getClass()) { // final implementation
   * },
   *
   * &nbsp;&nbsp;Names.named("mongodb")
   *
   * );
   */
  fun <From1 : DBComponent<*, *>, From2 : DBComponent<*, *>, From3 : From1, Target : From1> bind(
    from1: TypeToken<From1>,
    from2: TypeToken<From2>,
    from3: TypeToken<From3>,
    target: TypeToken<Target>,
    componentAnnotation: Annotation,
  )

  /**
   * Binding method for a component
   *
   *
   * A typical example of usage of this method:
   *
   * be.bind(
   *
   * &nbsp;&nbsp;new TypeToken<FooRepository<?, ?>>(getClass()) {
   * },
   *
   * &nbsp;&nbsp;new TypeToken<FooRepository<ObjectId, Datastore>>(getClass()) {
   * },
   *
   * &nbsp;&nbsp;new TypeToken<CommonMongoFooRepository>(getClass()) { // final implementation
   * },
   * &nbsp;&nbsp;Names.named("mongodb")
   *
   * );
   */
  fun <From1 : DBComponent<*, *>, From2 : From1, Target : From1> bind(
    from1: TypeToken<From1>,
    from2: TypeToken<From2>,
    target: TypeToken<Target>,
    componentAnnotation: Annotation,
  )

  fun <From, Target : From> bind(
    from: TypeToken<From>,
    target: TypeToken<Target>,
    annotation: Annotation,
  )

  fun <From, Target : From> bind(
    from: TypeToken<From>,
    target: TypeToken<Target>,
    annotation: Class<out Annotation>,
  )

  fun <From, Target : From> bind(
    from: TypeToken<From>,
    target: TypeToken<Target>,
  )

  /**
   * Binds the mongodb [DataStoreContext]
   *
   * Using this method is the same as invoking:
   *
   * bind(new TypeLiteral<DataStoreContext<ObjectId, Datastore>>() {
   * }).to(MongoContext.class);
   */
  fun withMongoDB()

  /**
   * Binds the xodus [DataStoreContext]
   *
   * Using this method is the same as invoking:
   *
   * binder.bind(new TypeLiteral<DataStoreContext<EntityId, PersistentEntityStore>>() {
   * }).to(XodusContext.class);
   *
   */
  fun withXodus()

  companion object {
    fun <T> getTypeLiteral(typeToken: TypeToken<T>): TypeLiteral<T> {
      return TypeLiteral.get(typeToken.type) as TypeLiteral<T>
    }

    fun <T> getKey(typeToken: TypeToken<T>): Key<T>? {
      return Key.get(getTypeLiteral(typeToken))
    }

    fun <T> asInternalProvider(clazz: Class<T>): Provider<T> {
      return Provider { Anvil.getEnvironment().injector.getInstance(clazz) }
    }

    fun <T> asInternalProvider(typeLiteral: TypeLiteral<T>): Provider<T> {
      return Anvil.getEnvironment().injector.getProvider(Key.get(typeLiteral))
    }

    fun <T> asInternalProvider(typeToken: TypeToken<T>): Provider<T> {
      return Anvil.getEnvironment().injector.getProvider(getKey(typeToken))
    }
  }
}
